### ISSUE TITLE
Segfault when root package is not defined

### ISSUE THREAD

**Issue by allen-servedio** [apache/incubator-openwhisk-wskdeploy#740]
I get the segfault shown below when a root package is not defined in my `manifest.yaml` for deploying dependent packages. 

For an example - this produces the segfault:

```
project:
  name: WskDeployWithDeps
  version: 0.0.2
  packages:
    myrediskv:
      license: Apache-2.0
      version: 0.0.2
      namespace: my-namespace
      dependencies:
        poc-common-func:
          location: https://github.com/my-org/my-project
          version: 0.0.1
          inputs: 
            redisURL:
              type: string
              value: my-url-goes-here
              description: Redis database URL
```

The following works properly because `deployers/servicedeployer.go:472` expects there to be a root package defined:

```
project:
  name: WskDeployWithDeps
  version: 0.0.2
  packages:
    myrediskv:
      license: Apache-2.0
      version: 0.0.2
      namespace: my-namespace
      dependencies:
        poc-common-func:
          location: https://github.com/my-org/my-project
          version: 0.0.1
          inputs: 
            redisURL:
              type: string
              value: my-url-goes-here
              description: Redis database URL
package:
  name: rediskv
```

Here is the segfault:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13bea4a]

goroutine 1 [running]:
github.com/apache/incubator-openwhisk-wskdeploy/deployers.(*ServiceDeployer).DeployDependencies(0xc420102240, 0x0, 0x0)
	/Users/aervedio/go/src/github.com/apache/incubator-openwhisk-wskdeploy/deployers/servicedeployer.go:472 +0xa4a
github.com/apache/incubator-openwhisk-wskdeploy/deployers.(*ServiceDeployer).deployAssets(0xc420102240, 0x0, 0x14c0677)
	/Users/aervedio/go/src/github.com/apache/incubator-openwhisk-wskdeploy/deployers/servicedeployer.go:374 +0x50
github.com/apache/incubator-openwhisk-wskdeploy/deployers.(*ServiceDeployer).Deploy(0xc420102240, 0x0, 0x0)
	/Users/aervedio/go/src/github.com/apache/incubator-openwhisk-wskdeploy/deployers/servicedeployer.go:339 +0x2c9
github.com/apache/incubator-openwhisk-wskdeploy/cmd.Deploy(0x0, 0x0)
	/Users/aervedio/go/src/github.com/apache/incubator-openwhisk-wskdeploy/cmd/root.go:243 +0x4c5
github.com/apache/incubator-openwhisk-wskdeploy/cmd.RootCmdImp(0x171fac0, 0xc4201fc340, 0x0, 0x4, 0x0, 0x0)
	/Users/aervedio/go/src/github.com/apache/incubator-openwhisk-wskdeploy/cmd/root.go:55 +0x22
github.com/spf13/cobra.(*Command).execute(0x171fac0, 0xc4200a0100, 0x4, 0x4, 0x171fac0, 0xc4200a0100)
	/Users/aervedio/go/src/github.com/spf13/cobra/command.go:753 +0x475
github.com/spf13/cobra.(*Command).ExecuteC(0x171fac0, 0x1a, 0x0, 0xc4200d1f70)
	/Users/aervedio/go/src/github.com/spf13/cobra/command.go:843 +0x334
github.com/spf13/cobra.(*Command).Execute(0x171fac0, 0x10037bf, 0x100371e)
	/Users/aervedio/go/src/github.com/spf13/cobra/command.go:791 +0x2b
github.com/apache/incubator-openwhisk-wskdeploy/cmd.Execute()
	/Users/aervedio/go/src/github.com/apache/incubator-openwhisk-wskdeploy/cmd/root.go:69 +0x40
main.main()
	/Users/aervedio/go/src/github.com/apache/incubator-openwhisk-wskdeploy/main.go:26 +0x20
```

The actual line that is failing is this one:

```
qName, err := utils.ParseQualifiedName(depServiceDeployer.RootPackageName, depServiceDeployer.Deployment.Packages[depServiceDeployer.RootPackageName].Package.Namespace)
```

I traced the code and can see that `depServiceDeployer.Deployment.Packages[depServiceDeployer.RootPackageName]` returns a nil (also `depServiceDeployer.RootPackageName` is empty) and so subsequent property dereferencing blows up with a nil pointer.

**Comment by allen-servedio** [comment]
And, now that I am looking closer, it looks like I am wrong about - at least - the work around. When I add the `package` definition to the bottom, the tool ignores the dependency and ONLY tries to deploy the package. So, while it no longer segfaults, it also does not deploy the dependencies I defined... The first config that segfaults at the end at least deploys the dependencies okay (and can undeploy them).

**Comment by mrutkows** [comment]
@allen-servedio is it possible to open granular issues for each case?  To me, it still seems like the non-deprecated, canonical schema (as described in spec.) works.  Then it is a matter of emitting the proper warnings and/or errors for those that follow deprecated (or incorrect) schema.

**Comment by allen-servedio** [comment]
Okay. I created #742 for it ignoring the packages section. So the scope of this defect is the segfault issue.

In case it helps, the `manifest.yaml` for the dependency looks like this:

```
project:
  name: PocCommonFunc
  version: 0.0.1
  packages:
    rediskv:
      license: Apache-2.0
      inputs: 
        redisURL:
          type: string
          description: Redis database URL
      actions: 
        kvget:
          function: retrieve/index.js
          runtime: nodejs
        kvput:
          function: store/index.js
          runtime: nodejs
```

**Comment by mrutkows** [comment]
@allen-servedio Running the first manifest (above) that causes you a segfault works for me and correctly reports on error (Unix/Mac):
```
./incubator-openwhisk-wskdeploy -i -v -m ~/manifest.yaml 
Info: Configuration:
      > WithinOpenWhisk: [false]
      > ApiHost: []
      > Auth: []
      > Namespace: []
      > ApiVersion: []
      > CfgFile: [/Users/Matt/.wskprops]
      > CliVersion: [unset]
      > CliBuild: [unset]
      > Verbose: [true]
      > ProjectPath: [.]
      > DeploymentPath: []
      > ManifestPath: [/Users/Matt/manifest.yaml]
      > UseDefaults: [false]
      > UseInteractive: [true]
      > Strict: [false]
      > Key: []
      > Cert: []
      > Managed: [false]

Info: The API host is [openwhisk.ng.bluemix.net], from .wskprops.
Info: The auth key is set, from .wskprops.
Info: The namespace is [_], from .wskprops.
[ploy/utils.ParseOpenWhisk]:116:[Inf] Unmarshal OpenWhisk info from internet.
Error: manifestreader.go [103]: [ERROR_YAML_FILE_FORMAT_ERROR]: File: [manifest.yaml]: 
==> manifestreader.go [147]: [ERROR_YAML_FILE_FORMAT_ERROR]: File: [poc-common-func]:
==> ==> zip: not a valid zip file
==> 

```

**Comment by allen-servedio** [comment]
Sorry, I should have specified, the code I am running is based on this pull request: https://github.com/apache/incubator-openwhisk-wskdeploy/pull/738

For you to get the segfault, it is going to have to point to a valid repo.

**Comment by mrutkows** [comment]
@allen-servedio  I am confused.  You submitted a PR to support dependencies on private repos that if we accept it results in an error elsewhere in the code?  Why does the PR not address this further issue as well as add a testcase for it?

**Comment by mrutkows** [comment]
@allen-servedio how do we automate a testcase for this (i.e., with someones private repo. credentials)?

**Comment by allen-servedio** [comment]
Okay, sorry @mrutkows for wasting your time with a terrible example. I have pushed a public repo that you can use to reproduce the error. I used the the version of wskdeploy that is in your current master branch.

Go ahead and do this:

```
git clone git@github.com:allen-servedio/wskdeploy-defect-740.git
cd wskdeploy-defect-740/defect-740
wskdeploy -v -m manifest.yaml
```

And you will get this failure:

```
Info: action [hellopkg/hello] has been successfully deployed.

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2deb87]

goroutine 1 [running]:
github.com/apache/incubator-openwhisk-wskdeploy/deployers.(*ServiceDeployer).DeployDependencies(0x10983730, 0x0, 0x0)
	/home/travis/gopath/src/github.com/apache/incubator-openwhisk-wskdeploy/deployers/servicedeployer.go:472 +0x8d7
github.com/apache/incubator-openwhisk-wskdeploy/deployers.(*ServiceDeployer).deployAssets(0x10983730, 0x0, 0x0)
	/home/travis/gopath/src/github.com/apache/incubator-openwhisk-wskdeploy/deployers/servicedeployer.go:374 +0x3b
github.com/apache/incubator-openwhisk-wskdeploy/deployers.(*ServiceDeployer).Deploy(0x10983730, 0x0, 0x0)
	/home/travis/gopath/src/github.com/apache/incubator-openwhisk-wskdeploy/deployers/servicedeployer.go:358 +0x3de
github.com/apache/incubator-openwhisk-wskdeploy/cmd.Deploy(0x1, 0x1097f7ac)
	/home/travis/gopath/src/github.com/apache/incubator-openwhisk-wskdeploy/cmd/root.go:241 +0x45d
github.com/apache/incubator-openwhisk-wskdeploy/cmd.RootCmdImp(0x565720, 0x109c9a40, 0x0, 0x3, 0x0, 0x0)
	/home/travis/gopath/src/github.com/apache/incubator-openwhisk-wskdeploy/cmd/root.go:54 +0x14
github.com/spf13/cobra.(*Command).execute(0x565720, 0x108ac388, 0x3, 0x3, 0x565720, 0x108ac388)
	/home/travis/gopath/src/github.com/spf13/cobra/command.go:632 +0x2df
github.com/spf13/cobra.(*Command).ExecuteC(0x565720, 0x3, 0x0, 0x0)
	/home/travis/gopath/src/github.com/spf13/cobra/command.go:722 +0x285
github.com/spf13/cobra.(*Command).Execute(0x565720, 0x0, 0x0)
	/home/travis/gopath/src/github.com/spf13/cobra/command.go:681 +0x1b
github.com/apache/incubator-openwhisk-wskdeploy/cmd.Execute()
	/home/travis/gopath/src/github.com/apache/incubator-openwhisk-wskdeploy/cmd/root.go:68 +0x30
main.main()
	/home/travis/gopath/src/github.com/apache/incubator-openwhisk-wskdeploy/main.go:26 +0x11
```

I then run `wsk package list` and `wsk action list` and can see that it actually deployed the `hellopkg` into my namespace and the `hellopkg/hello` action into it.

You can then use `wskdeploy undeploy -v -m manifest.yaml` to clean up your environment.

Next time I'll give you a better example - thank you for your patience.

**Comment by pritidesai** [comment]
Please refer to my previous comment, (fixed with PR #744), closing this.
